matgen/generate_matrix_he_and_tz.cc

```diff
@@ -0,0 +1,403 @@
+///
+template <typename scalar_t>
+void generate_matrix(
+    MatrixParams& params,
+    slate::HermitianMatrix<scalar_t>& A,
+    std::vector< blas::real_type<scalar_t> >& Sigma,
+    slate::Options const& opts)
+{
+    slate::BaseTrapezoidMatrix<scalar_t>& TZ = A;
+    generate_matrix( params, TZ, Sigma, opts );
+
+    // Set diagonal to real.
+    #pragma omp parallel for
+    for (int64_t i = 0; i < A.mt(); ++i) {
+        if (A.tileIsLocal( i, i )) {
+            A.tileGetForWriting( i, i, LayoutConvert::ColMajor );
+            auto T = A( i, i );
+            int64_t mb = T.mb();
+            for (int64_t ii = 0; ii < mb; ++ii) {
+                T.at( ii, ii ) = std::real( T( ii, ii ) );
+            }
+        }
+    }
+    A.tileUpdateAllOrigin();
+}
+
+// -----------------------------------------------------------------------------
+/// Generates an m-by-n trapezoid-storage test matrix.
+/// Handles Trapezoid, Triangular, Symmetric, and Hermitian matrices.
+/// @see generate_matrix
+/// @ingroup generate_matrix
+///
+template <typename scalar_t>
+void generate_matrix(
+    MatrixParams& params,
+    slate::BaseTrapezoidMatrix<scalar_t>& A,
+    std::vector< blas::real_type<scalar_t> >& Sigma,
+    slate::Options const& opts)
+{
+    using real_t = blas::real_type<scalar_t>;
+
+    // Constants
+    const real_t nan = std::numeric_limits<real_t>::quiet_NaN();
+    const real_t d_zero = 0;
+    const real_t d_one  = 1;
+    const scalar_t zero = 0;
+    const scalar_t one  = 1;
+
+    // ----------
+    // set Sigma to unknown (nan)
+    lapack::laset( lapack::MatrixType::General, Sigma.size(), 1,
+                   nan, nan, Sigma.data(), Sigma.size() );
+
+    char msg[ 256 ];
+    TestMatrixType type;
+    TestMatrixDist dist;
+    real_t cond;
+    real_t condD;
+    real_t sigma_max;
+    bool dominant;
+    int64_t zero_col;
+    decode_matrix<scalar_t>(
+        params, A, type, dist, cond, condD, sigma_max, dominant, zero_col );
+
+    int64_t seed = configure_seed(A.mpiComm(), params.seed());
+
+    int64_t n = A.n();
+    int64_t nt = A.nt();
+    int64_t mt = A.mt();
+
+    // ----- generate matrix
+    switch (type) {
+        case TestMatrixType::zeros:
+            set(zero, zero, A);
+            lapack::laset( lapack::MatrixType::General, Sigma.size(), 1,
+                d_zero, d_zero, Sigma.data(), Sigma.size() );
+            break;
+
+        case TestMatrixType::ones:
+            set(one, one, A);
+            lapack::laset( lapack::MatrixType::General, Sigma.size(), 1,
+                d_zero, d_one, Sigma.data(), Sigma.size() );
+            if (Sigma.size() >= 1) {
+                Sigma[0] = n;
+            }
+            break;
+
+        case TestMatrixType::identity:
+            set(zero, one, A);
+            lapack::laset( lapack::MatrixType::General, Sigma.size(), 1,
+                d_one, d_one, Sigma.data(), Sigma.size() );
+            break;
+
+        case TestMatrixType::jordan: {
```

Comment:
I have added this change to the `generate_matrix_ge.cc`; however, I still need to add the set function for `BaseTrapezoid` before making this change. Otherwise there is an error that occurs. I will work on that, from glancing at it, I believe it is just a copy and paste, but I will see.